## KMedoids
---

In [1]:
!pip install --upgrade --force-reinstall \
  "numpy==1.26.4" "scipy==1.11.4" "scikit-learn==1.3.2" "scikit-learn-extra==0.3.0"

import numpy as np, sklearn, sklearn_extra
import gower
from sklearn_extra.cluster import KMedoids
from sklearn.metrics import silhouette_score
print(np.__version__, sklearn.__version__)

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 61.0/61.0 kB 2.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 60.4/60.4 kB 3.4 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 819.0/819.0 kB 16.7 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 18.0/18.0 MB 78.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 35.8/35.8 MB 25.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 10.8/10.8 MB 101.1 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 308.4/308.4 kB 17.3 MB/s eta 0:00:00
  Created wheel for scikit-learn-extra: filename=scikit_learn_extra-0.3.0-cp312-cp312-linux_x86_64.whl size=2176928 sha256=f588c1b0bd3b6b98f81155f016e479e02e6624be31677e7a3a4c0c9993407eec
  Stored in directory: /root/.cache/pip/wheels/17/4d/c3/c6d5d563c1bf8146d059d63be3678abc2f2801fba0aaf5f0b8
Successful

ValueError: numpy.dtype size changed, may indicate binary incompatibility. Expected 96 from C header, got 88 from PyObject

### Descripcion de la base de datos

In [1]:
import pandas as pd
data = pd.read_csv('data_procesada.csv')
data.head()

AÑO     GPSLAT    GPSLONG  SECTOR  SECCION  MANZANA  DIRECCION_FILTRO  \
0  2018  10.952493 -74.777018    2201        8        3    Kr 14 # 21 105   
1  2018        NaN        NaN     203        1       13  kra 46A # 110_64   
2  2018        NaN        NaN     203        1       13  Kra 46A # 110_60   
3  2018   6.299582 -75.550590     203        2        2   kra 109 201 _28   
4  2018   7.090175 -73.145032     798        2       25    Cll 58 47 aw 8   

   DANE5  PB1  PERSONAS  ...  MUNICIPIO_NOMBRE  Nombre Municipio  \
0   8001    1         1  ...      Barranquilla      BARRANQUILLA   
1   5001    1         5  ...          Medellín          MEDELLÍN   
2   5001    1         4  ...          Medellín          MEDELLÍN   
3   5001    1         2  ...          Medellín          MEDELLÍN   
4  68001    1         4  ...       Bucaramanga       BUCARAMANGA   

   Nombre Departamento  Servicios_Telecomunicaciones_No  \
0            ATLÁNTICO                                1   
1            ANTIOQUIA                                0   
2            ANTIOQUIA                                0   
3            ANTIOQUIA                                0   
4            SANTANDER                                0   

   Servicios_Telecomunicaciones_Si Dispositivos_hogar_No  conexion_hogar_si  \
0                                0                  True              False   
1                                1                  True              False   
2                                1                  True              False   
3                                1                 False               True   
4                                1                 False               True   

   interrupciones_si  frec_uso_si  dens_int  
0              False         True  0.301608  
1              False         True  0.190441  
2              False         True  0.190441  
3              False         True  0.190441  
4               True         True  0.169252  

[5 rows x 37 columns]

In [2]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7662 entries, 0 to 7661
Data columns (total 37 columns):
 #   Column                           Non-Null Count  Dtype  
---  ------                           --------------  -----  
 0   AÑO                              7662 non-null   int64  
 1   GPSLAT                           6620 non-null   float64
 2   GPSLONG                          6620 non-null   float64
 3   SECTOR                           7662 non-null   int64  
 4   SECCION                          7662 non-null   int64  
 5   MANZANA                          7662 non-null   int64  
 6   DIRECCION_FILTRO                 7662 non-null   object 
 7   DANE5                            7662 non-null   int64  
 8   PB1                              7662 non-null   int64  
 9   PERSONAS                         7662 non-null   int64  
 10  EDAD                             7662 non-null   int64  
 11  REDAD                            7662 non-null   int64  
 12  GENERO              

In [3]:
data.describe().transpose()

count           mean            std  \
AÑO                              7662.0    2020.727095       2.489827   
GPSLAT                           6620.0       6.140382       3.417394   
GPSLONG                          6620.0       7.558627      74.597830   
SECTOR                           7662.0    1993.167189    3216.746402   
SECCION                          7662.0      12.033803      33.062799   
MANZANA                          7662.0      19.517750     117.912282   
DANE5                            7662.0   36881.520230   27746.655225   
PB1                              7662.0       1.111720       0.315042   
PERSONAS                         7662.0       3.521926       1.746042   
EDAD                             7662.0      40.037588      18.351039   
REDAD                            7662.0       3.978596       1.894850   
GENERO                           7662.0       1.506787       0.499987   
ESTRATO                          7662.0       2.271861       1.233094   
P9                               6683.0      62.729164     230.834651   
P34                              6522.0       1.387918       0.999507   
P56                              6791.0       7.137240       2.608756   
P57                               538.0       3.814126       1.465565   
P64                              6804.0       8.059083       2.057895   
NIVEL_PIRAMIDE                   7662.0       1.186635       0.948560   
POBLACION_5_16                   7662.0  259589.008353  338968.069308   
Estudiantes_5_16                 7662.0  244457.333072  317908.675196   
PROP_EDUC_5_16_MEN               7662.0       0.948225       0.100197   
HOGARES_INTERNET                 7662.0   14835.851475   22211.532355   
POBLACIÓN_ICFES                  7662.0   17478.427173   24521.054932   
TASA_INTERNET_ICFES              7662.0       0.732950       0.169146   
Servicios_Telecomunicaciones_No  7662.0       0.058862       0.235381   
Servicios_Telecomunicaciones_Si  7662.0       0.941138       0.235381   
dens_int                         7662.0       0.267050       0.169146   

                                         min           25%            50%  \
AÑO                              2018.000000   2018.000000    2023.000000   
GPSLAT                            -77.277221      4.432272       5.675870   
GPSLONG                           -77.297173    -74.826292      73.115492   
SECTOR                              0.000000     13.000000     705.000000   
SECCION                             0.000000      2.000000       3.000000   
MANZANA                             0.000000      4.000000       9.000000   
DANE5                            5001.000000   8758.000000   25843.000000   
PB1                                 1.000000      1.000000       1.000000   
PERSONAS                            1.000000      2.000000       3.000000   
EDAD                               12.000000     24.000000      38.000000   
REDAD                               1.000000      2.000000       4.000000   
GENERO                              1.000000      1.000000       2.000000   
ESTRATO                             0.000000      1.000000       2.000000   
P9                                  1.000000      2.000000       9.000000   
P34                                 1.000000      1.000000       1.000000   
P56                                 1.000000      5.000000       9.000000   
P57                                 1.000000      2.000000       5.000000   
P64                                 1.000000      9.000000       9.000000   
NIVEL_PIRAMIDE                      0.000000      0.000000       1.000000   
POBLACION_5_16                   1089.000000  40704.000000  103748.000000   
Estudiantes_5_16                 1337.000000  36043.000000   97241.000000   
PROP_EDUC_5_16_MEN                  0.650993      0.911592       0.965100   
HOGARES_INTERNET                   13.000000   1610.000000    5572.000000   
POBLACIÓN_ICFES                    53.000000   1895.000000    6995.0

### 1. Importar librerias

In [4]:
!pip install scikit-learn-extra

In [5]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import silhouette_score, calinski_harabasz_score, davies_bouldin_score

# ============================================
# IMPORTAR K-MEDOIDS
# ============================================
from sklearn_extra.cluster import KMedoids

### 2. Preparación de Datos

In [6]:
FEATURES = [
    "Servicios_Telecomunicaciones_No", "Servicios_Telecomunicaciones_Si",
    "Dispositivos_hogar_No", "conexion_hogar_si", "interrupciones_si",
    "frec_uso_si", "dens_int",
]

def load_data(path="data_procesada.csv"):
    df = pd.read_csv(path)
    for c in FEATURES:
        if c != "dens_int":
            df[c] = df[c].astype("float").astype("Int64").fillna(0).astype(int)
        else:
            df[c] = pd.to_numeric(df[c], errors="coerce")
    if df["dens_int"].isna().any():
        df["dens_int"] = df["dens_int"].fillna(df["dens_int"].median())
    return df

def apply_weights(X: pd.DataFrame, weights: dict | None):
    Xw = X.copy()
    if weights:
        for col, w in weights.items():
            if col in Xw.columns:
                Xw[col] = Xw[col].astype(float) * float(w)
    return Xw

def scale_numeric(X: pd.DataFrame, numeric_cols=("dens_int",)):
    Xs = X.copy()
    scaler = StandardScaler()
    Xs[list(numeric_cols)] = scaler.fit_transform(Xs[list(numeric_cols)])
    return Xs

# Cargamos los datos
df = load_data("data_procesada.csv")

### 3. Función para K-Medoids

In [7]:
from sklearn.metrics import silhouette_score, calinski_harabasz_score, davies_bouldin_score, adjusted_rand_score

def compute_internal_metrics(X: np.ndarray, labels: np.ndarray) -> dict:

    if len(np.unique(labels)) > 1:
        sil = silhouette_score(X, labels)
        ch = calinski_harabasz_score(X, labels)
        db = davies_bouldin_score(X, labels)
    else:
        sil, ch, db = np.nan, np.nan, np.nan
    return {"silhouette": sil, "calinski_harabasz": ch, "davies_bouldin": db}

def profile_sharpness(centroids: pd.DataFrame, binary_cols: list[str]) -> float:
    """Promedio de distancia del centroide a 0.5 (cuanto mayor, más 'nítido')."""
    if not binary_cols: return np.nan
    diffs = [np.abs(centroids[col] - 0.5) for col in binary_cols if col in centroids.columns]
    if not diffs: return np.nan
    return float(pd.concat(diffs, axis=1).mean().mean())

def cluster_size_ok(labels: np.ndarray, min_ratio=0.02) -> bool:
    n = len(labels)
    sizes = pd.Series(labels).value_counts(normalize=True)
    return (sizes >= min_ratio).all()

def centroid_table(X: pd.DataFrame, labels: np.ndarray):
    """Promedio por clúster (binarios → tasas; dens_int → media)."""
    dfc = X.copy()
    dfc["cluster"] = labels
    cent = dfc.groupby("cluster").mean(numeric_only=True).sort_index()
    sizes = dfc["cluster"].value_counts().sort_index().to_frame(name="n")
    sizes["pct"] = sizes["n"] / len(X)
    return cent, sizes


def stability_ari(X: np.ndarray, k: int, seeds: list[int], subsample=0.8) -> float:
    """Promedio de ARI entre múltiples corridas con submuestreo."""
    from sklearn_extra.cluster import KMedoids

    label_list = []
    rng = np.random.default_rng(42)
    n = X.shape[0]
    for seed in seeds:
        idx = rng.choice(n, size=int(n*subsample), replace=False)

        model = KMedoids(n_clusters=k, random_state=seed)
        labels = model.fit_predict(X[idx])
        label_list.append((idx, labels))

    if len(label_list) < 2: return np.nan
    aris = []
    for i in range(len(label_list)):
        for j in range(i+1, len(label_list)):
            idx_i, lab_i = label_list[i]
            idx_j, lab_j = label_list[j]
            common_idx = np.intersect1d(idx_i, idx_j, assume_unique=True)
            if len(common_idx) < 10: continue


            map_i = {val: pos for pos, val in enumerate(idx_i)}
            map_j = {val: pos for pos, val in enumerate(idx_j)}

            li = [lab_i[map_i[val]] for val in common_idx]
            lj = [lab_j[map_j[val]] for val in common_idx]

            aris.append(adjusted_rand_score(li, lj))
    return float(np.mean(aris)) if aris else np.nan

def geographic_entropy(df: pd.DataFrame, labels: np.ndarray, col_region="Nombre Departamento") -> float:
    """Entropía normalizada promedio de la distribución regional por clúster."""
    df_ = df[[col_region]].copy()
    df_["cluster"] = labels
    entropies = []
    for c, g in df_.groupby("cluster"):
        p = g[col_region].value_counts(normalize=True)
        H = -(p * np.log(p + 1e-12)).sum()
        H_max = np.log(len(p)) if len(p) > 1 else 1.0
        entropies.append(H / H_max)
    return float(np.mean(entropies)) if entropies else np.nan

In [8]:
# FUNCIÓN PARA K-MEDOIDS
def run_kmedoids_grid(df: pd.DataFrame, k_values, weights=None, exp_name="exp"):
    # Preparación de datos
    X = df[FEATURES].copy()
    X = apply_weights(X, weights)
    X = scale_numeric(X, numeric_cols=("dens_int",))
    X_np = X.values

    results = []
    best = None

    for k in k_values:
        # Usa k-medoids
        km = KMedoids(n_clusters=k, init="k-medoids++", random_state=123)
        labels = km.fit_predict(X_np)

        m = compute_internal_metrics(X_np, labels)
        ok = cluster_size_ok(labels, min_ratio=0.02)

        # Los centroides en K-Medoids son los medoides (puntos reales)
        # La función centroid_table calcula promedios para perfilar el clúster
        cent, sizes = centroid_table(X, labels)
        ps = profile_sharpness(cent, [c for c in FEATURES if c != "dens_int"])

        # Se mide la estabilidad
        ari = stability_ari(X_np, k, seeds=[0,1,2,3,4], subsample=0.8)

        row = {
            "exp": exp_name, "k": k, "silhouette": m["silhouette"],
            "calinski_harabasz": m["calinski_harabasz"], "davies_bouldin": m["davies_bouldin"],
            "profile_sharpness": ps, "stability_ari": ari,
            "min_size_ok": ok, "sizes": sizes["pct"].round(3).to_dict()
        }
        results.append(row)

        # Lógica de selección del mejor clúster
        if ok:
            if best is None or (row["silhouette"] > best[0]["silhouette"]):
                best = (row, labels, cent, sizes)

    results_df = pd.DataFrame(results).sort_values("silhouette", ascending=False)
    return results_df, best

### 4. Ejecución de los Experimentos con K-Medoids

In [9]:
# Experimento 1 — Línea base con K-Medoids

print("--- EJECUTANDO EXPERIMENTO 1 CON K-MEDOIDS ---")
k_values_exp1 = [4, 5, 6, 7, 8, 9, 10, 12]
res1_kmedoids, best1_kmedoids = run_kmedoids_grid(df, k_values_exp1, weights=None, exp_name="exp1_kmedoids")

print("Resumen EXP1 (K-Medoids):")
print(res1_kmedoids.head())
print("\nMejor resultado EXP1 (K-Medoids):", best1_kmedoids[0])


# Experimento 2 — Ponderación con K-Medoids

print("\n--- EJECUTANDO EXPERIMENTO 2 CON K-MEDOIDS ---")
weights_exp2 = {
    "Servicios_Telecomunicaciones_Si": 2.0,
    "conexion_hogar_si": 2.0,
    "frec_uso_si": 2.0, # (Asegúrate que esta columna exista o ajústala)
    "dens_int": 2.0
}
k_values_exp2 = [4, 5, 6, 7, 8]
res2_kmedoids, best2_kmedoids = run_kmedoids_grid(df, k_values_exp2, weights=weights_exp2, exp_name="exp2_kmedoids")

print("Resumen EXP2 (K-Medoids):")
print(res2_kmedoids.head())
print("\nMejor resultado EXP2 (K-Medoids):", best2_kmedoids[0])

# Experimento 3 — Estabilidad y Coherencia (con K-Medoids)

# Recuperar mejores resultados
# ('best1_kmedoids' y 'best2_kmedoids')
row1_kmedoids, labels1_kmedoids, cent1_kmedoids, sizes1_kmedoids = best1_kmedoids
row2_kmedoids, labels2_kmedoids, cent2_kmedoids, sizes2_kmedoids = best2_kmedoids


##  Lógica de selección del mejor modelo
def prefer(a, b):
    # Criterio 1: Silhouette
    if a["silhouette"] > b["silhouette"] + 1e-6: return "A"
    if b["silhouette"] > a["silhouette"] + 1e-6: return "B"

    # Criterio 2 Estabilidad ARI
    if a["stability_ari"] > b["stability_ari"] + 1e-6: return "A"
    if b["stability_ari"] > a["stability_ari"] + 1e-6: return "B"

    # Criterio 3 Nitidez de Perfil
    if a["profile_sharpness"] >= b["profile_sharpness"]: return "A"
    return "B"

## Comparar los mejores de cada experimento de K-Medoids
choice = prefer(row1_kmedoids, row2_kmedoids)

# Preparación de la configuración ganadora

if choice == "A":
    print("\n[EXP3] Partimos de la mejor configuración de EXP1 (K-Medoids)")
    # Recrear los datos usados en el Exp 1
    X_choice = scale_numeric(apply_weights(df[FEATURES], None), ("dens_int",))
    k_choice = row1_kmedoids["k"]
    labels_choice = labels1_kmedoids
    exp_winner_name = "EXP1"
else:
    print("\n[EXP3] Partimos de la mejor configuración de EXP2 (K-Medoids)")
    # Recrear los datos usados en el Exp 2 (con pesos)
    X_choice = scale_numeric(apply_weights(df[FEATURES], weights_exp2), ("dens_int",))
    k_choice = row2_kmedoids["k"]
    labels_choice = labels2_kmedoids
    exp_winner_name = "EXP2"

print(f"Configuración ganadora: {exp_winner_name} con k = {k_choice}")

## Validación final de estabilidad y coherencia
# Ejecutar una prueba de estabilidad más exhaustiva con 30 repeticiones

print("\nEjecutando validación final de estabilidad y coherencia...")
ari30 = stability_ari(X_choice.values, k_choice, seeds=list(range(30)), subsample=0.8)
geoH3 = geographic_entropy(df, labels_choice, col_region="Nombre Departamento")

print(f"Estabilidad (ARI promedio con 30 seeds): {ari30:.3f}")
print(f"Coherencia Territorial (Entropía Geográfica promedio): {geoH3:.3f}")

# Reporte final de perfiles y tamaños
# Perfiles finales (promedios por clúster) y los tamaños relativos

centF, sizesF = centroid_table(df[FEATURES], labels_choice)

print("\n--- PERFILES FINALES (MEDOIDES) ---")
print(centF.round(3))

print("\n--- TAMAÑOS FINALES DE CLÚSTERS ---")
print(sizesF.assign(pct = (sizesF["pct"]*100).round(1)))

--- EJECUTANDO EXPERIMENTO 1 CON K-MEDOIDS ---


/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 1 is empty! self.labels_[self.medoid_indices_[1]] may not be labeled with its corresponding cluster (1).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 2 is empty! self.labels_[self.medoid_indices_[2]] may not be labeled with its corresponding cluster (2).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 3 is empty! self.labels_[self.medoid_indices_[3]] may not be labeled with its corresponding cluster (3).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 1 is empty! self.labels_[self.medoid_indices_[1]] may not be labeled with its corresponding cluster (1).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 2 

Resumen EXP1 (K-Medoids):
             exp   k  silhouette  calinski_harabasz  davies_bouldin  \
7  exp1_kmedoids  12    0.528031        4821.523390        0.898200   
6  exp1_kmedoids  10    0.523293        4968.102581        0.804909   
5  exp1_kmedoids   9    0.499269        4408.774786        0.900413   
4  exp1_kmedoids   8    0.482635        4343.167068        0.963007   
3  exp1_kmedoids   7    0.462061        4471.579529        0.997431   

   profile_sharpness  stability_ari  min_size_ok  \
7           0.462563            1.0        False   
6           0.461641            1.0         True   
5           0.445894            1.0         True   
4           0.441600            1.0         True   
3           0.436049            1.0         True   

                                               sizes  
7  {0: 0.236, 1: 0.059, 2: 0.11, 3: 0.052, 4: 0.1...  
6  {0: 0.236, 1: 0.066, 2: 0.123, 3: 0.064, 4: 0....  
5  {0: 0.236, 1: 0.066, 2: 0.128, 3: 0.083, 4: 0....  
4  {0: 0.236, 

/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 1 is empty! self.labels_[self.medoid_indices_[1]] may not be labeled with its corresponding cluster (1).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 2 is empty! self.labels_[self.medoid_indices_[2]] may not be labeled with its corresponding cluster (2).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 3 is empty! self.labels_[self.medoid_indices_[3]] may not be labeled with its corresponding cluster (3).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 1 is empty! self.labels_[self.medoid_indices_[1]] may not be labeled with its corresponding cluster (1).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 2 

Resumen EXP2 (K-Medoids):
             exp  k  silhouette  calinski_harabasz  davies_bouldin  \
4  exp2_kmedoids  8    0.550027        6202.963743        0.724845   
3  exp2_kmedoids  7    0.521519        5631.316283        0.765664   
0  exp2_kmedoids  4    0.457741        4635.032357        0.924421   
1  exp2_kmedoids  5    0.438175        4676.046230        0.885755   
2  exp2_kmedoids  6    0.404529        4979.212581        0.953644   

   profile_sharpness  stability_ari  min_size_ok  \
4           0.873669            1.0         True   
3           0.866689            1.0         True   
0           0.803552            1.0         True   
1           0.799905            1.0         True   
2           0.827014            1.0         True   

                                               sizes  
4  {0: 0.234, 1: 0.098, 2: 0.131, 3: 0.076, 4: 0....  
3  {0: 0.234, 1: 0.143, 2: 0.123, 3: 0.048, 4: 0....  
0           {0: 0.527, 1: 0.225, 2: 0.197, 3: 0.051}  
1  {0: 0.492, 1: 0.2

/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 1 is empty! self.labels_[self.medoid_indices_[1]] may not be labeled with its corresponding cluster (1).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 2 is empty! self.labels_[self.medoid_indices_[2]] may not be labeled with its corresponding cluster (2).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 3 is empty! self.labels_[self.medoid_indices_[3]] may not be labeled with its corresponding cluster (3).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 4 is empty! self.labels_[self.medoid_indices_[4]] may not be labeled with its corresponding cluster (4).
  warnings.warn(
/usr/local/lib/python3.12/dist-packages/sklearn_extra/cluster/_k_medoids.py:329: UserWarning: Cluster 5 

Estabilidad (ARI promedio con 30 seeds): 1.000
Coherencia Territorial (Entropía Geográfica promedio): 0.838

--- PERFILES FINALES (MEDOIDES) ---
         Servicios_Telecomunicaciones_No  Servicios_Telecomunicaciones_Si  \
cluster                                                                     
0                                  0.009                            0.991   
1                                  0.009                            0.991   
2                                  0.000                            1.000   
3                                  0.032                            0.968   
4                                  0.012                            0.988   
5                                  1.000                            0.000   
6                                  0.019                            0.981   
7                                  0.019                            0.981   

         Dispositivos_hogar_No  conexion_hogar_si  interrupciones_si  \
cluster     

### 5. Resumen de resultados

In [10]:
# La configuración ganadora fue la del modelo ponderado con K-Medoids
configuracion_ganadora = best2_kmedoids

metricas_finales = configuracion_ganadora[0]

metricas_finales['min_size_ok'] = bool(metricas_finales['min_size_ok'])

# El primer elemento de la tupla es el diccionario con las métricas
# Imprime las métricas de forma ordenada
import json
print(json.dumps(metricas_finales, indent=4))

{
    "exp": "exp2_kmedoids",
    "k": 8,
    "silhouette": 0.5500273081779461,
    "calinski_harabasz": 6202.963742728442,
    "davies_bouldin": 0.7248453817177823,
    "profile_sharpness": 0.8736685926362187,
    "stability_ari": 1.0,
    "min_size_ok": true,
    "sizes": {
        "0": 0.234,
        "1": 0.098,
        "2": 0.131,
        "3": 0.076,
        "4": 0.179,
        "5": 0.046,
        "6": 0.054,
        "7": 0.181
    }
}


In [11]:
row, labels, cent_scaled, sizes = best2_kmedoids  # tupla: (métricas, etiquetas, "centroides", tamaños)
metricas = pd.Series(row).to_frame("valor")
print(metricas)

# Tamaños de clúster (n y %)
print("\nTamaños de clúster:")
print(sizes.assign(pct=(sizes["pct"]*100).round(1)))

# Perfiles interpretables (en la escala original, NO ponderada/estandarizada)
#    — promedios de 0/1 (tasas) y media de dens_int
from pprint import pprint
centF, sizesF = centroid_table(df[FEATURES], labels)
print("\nPerfiles por clúster (escala original):")
print(centF.round(3))

# Silhouette por clúster (útil para ver qué grupos separan mejor)
from sklearn.metrics import silhouette_samples
X_choice = scale_numeric(apply_weights(df[FEATURES], {
    "Servicios_Telecomunicaciones_Si": 2.0,
    "conexion_hogar_si": 2.0,
    "frec_uso_si": 2.0,
    "dens_int": 2.0
}), ("dens_int",)).values

sil_s = silhouette_samples(X_choice, labels)
sil_por_cluster = pd.DataFrame({"cluster": labels, "sil": sil_s}).groupby("cluster").sil.mean().sort_values(ascending=False)
print("\nSilhouette medio por clúster:")
print(sil_por_cluster.round(3))

# Identificar los medoides y ver sus registros representativos
from sklearn_extra.cluster import KMedoids
km = KMedoids(n_clusters=row["k"], init="k-medoids++", random_state=123).fit(X_choice)
medoid_idx = km.medoid_indices_
cols_ver = ["Nombre Municipio", "Nombre Departamento"] + FEATURES
print("\nRegistros medoides (representantes) por clúster:")
print(df.iloc[medoid_idx][cols_ver])


                                                               valor
exp                                                    exp2_kmedoids
k                                                                  8
silhouette                                                  0.550027
calinski_harabasz                                        6202.963743
davies_bouldin                                              0.724845
profile_sharpness                                           0.873669
stability_ari                                                    1.0
min_size_ok                                                     True
sizes              {0: 0.234, 1: 0.098, 2: 0.131, 3: 0.076, 4: 0....

Tamaños de clúster:
            n   pct
cluster            
0        1794  23.4
1         752   9.8
2        1004  13.1
3         586   7.6
4        1368  17.9
5         356   4.6
6         417   5.4
7        1385  18.1

Perfiles por clúster (escala original):
         Servicios_Telecomunicaciones_No  Servic

* **Clúster 5 “Sin servicios de telecomunicaciones” (4,6%)**
Es el grupo más crítico: 100% sin servicios de telecomunicaciones y alta carencia de dispositivos (0,694). La conexión en el hogar es prácticamente nula (0,045) y la densidad de internet del entorno es baja-media (0,273). Prioridad alta para conectividad básica y dotación.

* **Clústeres 2 y 3 — “Sin conexión en el hogar” (13,1% y 7,6%)**
Ambos no tienen conexión doméstica (0,000) y muestran carencia relevante de dispositivos (0,577 y 0,544). Se diferencian por el contexto: el clúster 3 opera en zonas de alta densidad (0,547) frente a baja-media en el clúster 2 (0,201), lo que sugiere que en el clúster 3 las barreras son más de adopción/asequibilidad/uso que de oferta.

* **Clústeres 0 y 7 — “Conectados con mala calidad” (23,4% y 18,1%)**
Conexión en el hogar universal (1,000) y equipamiento disponible (0,000 sin dispositivos), pero interrupciones generalizadas (1,000). Densidades bajas-medias (0,126 y 0,308). Representan ~41,5% de la muestra: foco claro para acciones de mejora de calidad y confiabilidad del servicio.

* **Clústeres 1 y 4 — “Conectados estables/satisfechos” (9,8% y 17,9%)**
Conexión doméstica (1,000) sin interrupciones (0,000) y con dispositivos disponibles (0,000 sin dispositivos). Se distinguen por densidad: media en el clúster 1 (0,399) y baja en el clúster 4 (0,155). Constituyen la base “saludable” del ecosistema.

* **Clúster 6 — “Conectados intermitentes” (5,4%)**
Conectados (1,000) con interrupciones moderadas (0,571) y en el entorno de mayor densidad (0,627). El patrón sugiere congestión o problemas de capacidad: candidato a intervenciones de ampliación/gestión de red.